nats.go

```diff
@@ -2026,33 +2018,32 @@ func (nc *Conn) Request(subj string, data []byte, timeout time.Duration) (*Msg,
 	}
 
 	// snapshot
-	var doSetup, useOldRequestStyle bool
 	nc.mu.Lock()
-	useOldRequestStyle = nc.Opts.UseOldRequestStyle
-	doSetup = (nc.respMux == nil)
+	useOldRequestStyle := nc.Opts.UseOldRequestStyle
 	nc.mu.Unlock()
 
 	// If user wants the old style.
 	if useOldRequestStyle {
 		return nc.oldRequest(subj, data, timeout)
 	}
 
-	// Make sure scoped subscription is setup at least once on first
-	// call to Request(). Will handle duplicates in createRespMux.
-	if doSetup {
-		if err := nc.createRespMux(); err != nil {
-			return nil, err
-		}
+	// Make sure scoped subscription is setup only once on first call to
+	// Request().
+	var err error
+	nc.respSetup.Do(func() { err = nc.createRespMux() })
```

Comment:
All calls will block until `createRespMux()` returns, right? That is, say "thread" 1 is first to go, is in the middle of executing `createRespMux()` and a second "thread" executes `nc.respSetup.Do()`, then the call will block, right? If not, then the code would not be correct.

The only drawback with this (and previous) approach is that we lock nc, use now this lock (I am sure it is implemented with locking/atomic), then again lock nc to create the respInbox. In the C client, I initialize map and respInbox in the lock above, then *only* when we need to create the sub there is extra locking required. So after initial creation of subscription, I require only 1 lock, here it is 3 per request. That's fine, we can iterate again later.